# printing_client.py
import grpc
import time
import random
import argparse
import threading
from concurrent import futures
import distributed_printing_pb2 as pb
import distributed_printing_pb2_grpc as rpc

# ===========================
# Lamport clock (thread-safe)
# ===========================
lamport = 0
_lamport_lock = threading.Lock()

def increment():
    global lamport
    with _lamport_lock:
        lamport += 1
        return lamport

def update(received_ts):
    global lamport
    with _lamport_lock:
        lamport = max(lamport, received_ts) + 1
        return lamport

# =================================
# Servidor gRPC para cliente-cliente
# =================================
class MutualExclusionService(rpc.MutualExclusionServiceServicer):
    def __init__(self, client_obj):
        self.client = client_obj

    # Handler para RequestAccess (bloqueante): retorna AccessResponse quando concede
    def RequestAccess(self, request, context):
        # Atualiza rel√≥gio local ao receber
        update(request.lamport_timestamp)
        print(f"üì® [Cliente {self.client.id}] RequestAccess recebido de {request.client_id} (ts={request.lamport_timestamp})")
        # Implementa√ß√£o bloqueante: se eu estou pedindo e tenho prioridade, o servidor espera at√© liberar
        with self.client.lock:
            # Enquanto eu estiver requisitando e eu tiver prioridade sobre o requester, aguarda
            while self.client.requesting and (self.client.request_ts, self.client.id) < (request.lamport_timestamp, request.client_id):
                # aguardando at√© ser notificado por ReleaseAccess
                print(f"‚è≥ [Cliente {self.client.id}] Deferindo pedido de {request.client_id} (meu ts={self.client.request_ts})")
                self.client.wait_cv.wait()
            # conceder acesso (retorna imediatamente)
            # atualizar lamport antes de responder
            t = increment()
            print(f"‚úÖ [Cliente {self.client.id}] Concedendo acesso a {request.client_id} (resposta ts={t})")
            return pb.AccessResponse(ack=True, lamport_timestamp=t)

    # Handler para ReleaseAccess: notifica blocked request handlers para reavaliar
    def ReleaseAccess(self, request, context):
        update(request.lamport_timestamp)
        print(f"üîî [Cliente {self.client.id}] ReleaseAccess recebido de {request.client_id} (ts={request.lamport_timestamp})")
        with self.client.lock:
            # notificar possiveis handlers bloqueados
            self.client.wait_cv.notify_all()
        return pb.EmptyResponse()

class PrintingClient:
    def __init__(self, id, port, printer_addr, peers):
        """
        peers: lista de strings 'host:port' dos outros clientes (n√£o inclui o pr√≥prio cliente).
        """
        self.id = id
        self.port = port
        self.printer_addr = printer_addr
        self.peers = peers[:]  # lista de "host:port"
        # Estado do algoritmo
        self.requesting = False
        self.request_ts = None
        self.lock = threading.Lock()
        self.wait_cv = threading.Condition(self.lock)  # usado para handlers bloqueados e wait/notify
        # para contagem de acks (quando chamamos RequestAccess em peers)
        # n√≥s adotamos RequestAccess bloqueante: cada call retorna ack quando concedido
        # dessa forma, s√≥ precisamos contar quantas respostas recebemos (chamada retorna)
        self.running = True

    # ============ servidor cliente-cliente ============
    def start_server(self):
        server = grpc.server(futures.ThreadPoolExecutor(max_workers=20))
        rpc.add_MutualExclusionServiceServicer_to_server(MutualExclusionService(self), server)
        server.add_insecure_port(f"[::]:{self.port}")
        server.start()
        print(f"üõ∞Ô∏è [Cliente {self.id}] Servidor gRPC ativo na porta {self.port}")
        threading.Thread(target=server.wait_for_termination, daemon=True).start()

    # ============ Comunica√ß√£o com peers usando RequestAccess (bloqueante) ============
    def send_request_to_peers(self, ts):
        """
        Envia RequestAccess para todos os peers. Como a vers√£o que implementamos √© bloqueante,
        cada stub.RequestAccess(msg) s√≥ retorna quando o peer concede (ou d√° erro/timeout).
        """
        responses = 0
        for peer in self.peers:
            try:
                channel = grpc.insecure_channel(peer)
                stub = rpc.MutualExclusionServiceStub(channel)
                msg = pb.AccessRequest(
                    client_id=self.id,
                    lamport_timestamp=ts,
                    request_number=0
                )
                # Chamada bloqueante: retorna quando o peer concede
                # Colocamos timeout para n√£o travar indefinidamente (p.ex.: 10s)
                resp = stub.RequestAccess(msg, timeout=10)
                # Atualiza local Lamport com o timestamp que recebemos em resp (se houver)
                try:
                    update(resp.lamport_timestamp)
                except Exception:
                    pass
                if getattr(resp, "ack", False):
                    responses += 1
                    print(f"üì• [Cliente {self.id}] Ack recebido de {peer}")
                else:
                    print(f"‚ö†Ô∏è [Cliente {self.id}] Peer {peer} respondeu sem ack (poss√≠vel pol√≠tica).")
            except grpc.RpcError as e:
                print(f"‚ùå [Cliente {self.id}] Falha RequestAccess em {peer}: {e}")
        return responses

    def send_release_to_peers(self):
        """
        Notifica peers que eu liberei (ReleaseAccess). Isso faz com que os servidores dos peers
        notifiquem handlers bloqueados e reavaliem pedidos.
        """
        # incrementa lamport antes de enviar release
        ts = increment()
        for peer in self.peers:
            try:
                channel = grpc.insecure_channel(peer)
                stub = rpc.MutualExclusionServiceStub(channel)
                msg = pb.AccessRelease(
                    client_id=self.id,
                    lamport_timestamp=ts,
                    request_number=0
                )
                stub.ReleaseAccess(msg, timeout=5)
                # n√£o esperamos resposta al√©m do retorno do RPC
            except grpc.RpcError as e:
                print(f"‚ùå [Cliente {self.id}] Erro ao enviar Release para {peer}: {e}")

    # ============ Comunica√ß√£o com o servidor burro ============
    def send_to_printer(self, message):
        try:
            # incrementa lamport pois est√° enviando mensagem externa (regra pr√°tica)
            ts = increment()
            channel = grpc.insecure_channel(self.printer_addr)
            stub = rpc.PrintingServiceStub(channel)
            msg = pb.PrintRequest(
                client_id=self.id,
                message_content=message,
                lamport_timestamp=ts,
                request_number=0
            )
            response = stub.SendToPrinter(msg, timeout=10)
            # atualiza com timestamp retornado
            try:
                update(response.lamport_timestamp)
            except Exception:
                pass
            print(f"üñ®Ô∏è [Cliente {self.id}] Resposta da impressora: {response.confirmation_message}")
        except grpc.RpcError as e:
            print(f"‚ùå [Cliente {self.id}] Falha ao enviar para impressora: {e}")

    # ============ Se√ß√£o Cr√≠tica ============
    def critical_section(self):
        print(f"‚úÖ [Cliente {self.id}] Entrou na se√ß√£o cr√≠tica.")
        # Exemplo de conte√∫do da impress√£o: pode ser modificado para incluir request_ts, etc.
        self.send_to_printer(f"Documento do cliente {self.id} (ts={self.request_ts})")
        # simula trabalho na se√ß√£o cr√≠tica
        time.sleep(1)
        print(f"üö™ [Cliente {self.id}] Saindo da se√ß√£o cr√≠tica.")
        # libera e notifica peers
        with self.lock:
            self.requesting = False
            self.request_ts = None
            # notifica handlers locais que possam estar aguardando
            self.wait_cv.notify_all()
        # envia Release para peers
        self.send_release_to_peers()

    # ============ Loop Principal ============
    def run(self):
        print(f"üèÉ [Cliente {self.id}] Iniciando loop principal. Peers: {self.peers}")
        while self.running:
            # espera aleat√≥ria entre tentativas de entrar na se√ß√£o cr√≠tica
            time.sleep(random.randint(5, 10))
            # preparar pedido
            with self.lock:
                self.requesting = True
                ts = increment()
                self.request_ts = ts
                print(f"üÜò [Cliente {self.id}] Solicitando acesso (ts={ts})")
            # enviar RequestAccess para todos os peers (blocking calls)
            responses = self.send_request_to_peers(ts)
            # se conseguimos ack de todos os peers (respostas = len(peers)), entra na se√ß√£o cr√≠tica
            if responses >= len(self.peers):
                # entrou na se√ß√£o cr√≠tica
                self.critical_section()
            else:
                # caso algum peer n√£o tenha respondido, liberar estado e tentar depois
                print(f"‚ö†Ô∏è [Cliente {self.id}] N√£o obteve ack de todos os peers ({responses}/{len(self.peers)}). Abortando pedido e tentando depois.")
                with self.lock:
                    self.requesting = False
                    self.request_ts = None
                    self.wait_cv.notify_all()
                # opcional: esperar um tempo antes de tentar novamente
                time.sleep(1)

    def stop(self):
        self.running = False

# ===========================
# Execu√ß√£o principal
# ===========================
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--id", type=int, required=True, help="ID do cliente (inteiro √∫nico)")
    parser.add_argument("--server", type=str, required=True, help="Endere√ßo do servidor de impress√£o (host:port)")
    parser.add_argument("--port", type=int, required=True, help="Porta em que este cliente serve (ex: 50052)")
    parser.add_argument("--clients", type=str, required=True,
                        help="Lista de peers separada por v√≠rgula (ex: localhost:50053,localhost:50054). N√ÉO inclua este pr√≥prio cliente.")
    args = parser.parse_args()

    peers = [p.strip() for p in args.clients.split(",") if p.strip()]
    client = PrintingClient(args.id, args.port, args.server, peers)
    client.start_server()
    try:
        client.run()
    except KeyboardInterrupt:
        print("\nInterrompido pelo usu√°rio. Encerrando...")
        client.stop()
